src/local-ssh/sshTunnel.ts

```diff
@@ -0,0 +1,112 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Gitpod. All rights reserved.
+ *  Licensed under the MIT License. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+import { ChannelOpenMessage, SshClientCredentials, SshClientSession, SshDataWriter, SshSessionConfiguration, SshStream, Stream, WebSocketStream } from '@microsoft/dev-tunnels-ssh';
+import { TunnelPortRequest } from '@gitpod/supervisor-api-grpc/lib/port_pb';
+import { WebSocket } from 'ws';
+import { ControlService } from '@gitpod/supervisor-api-grpcweb/lib/control_pb_service';
+import { CreateSSHKeyPairRequest, CreateSSHKeyPairResponse } from '@gitpod/supervisor-api-grpcweb/lib/control_pb';
+import { NodeHttpTransport } from '@improbable-eng/grpc-web-node-http-transport';
+import { grpc } from '@improbable-eng/grpc-web';
+import { BrowserHeaders } from 'browser-headers';
+import { ILogger, WorkspaceAuthInfo } from './common';
+import { ConnectConfig } from 'ssh2';
+
+grpc.setDefaultTransport(NodeHttpTransport());
+
+class SupervisorPortTunnelMessage extends ChannelOpenMessage {
+	constructor(private clientId: string, private remotePort: number, channelType: string) {
+		super();
+		this.channelType = channelType;
+	}
+
+	override onWrite(writer: SshDataWriter): void {
+		super.onWrite(writer);
+		const req = new TunnelPortRequest();
+		req.setClientId(this.clientId);
+		req.setTargetPort(this.remotePort);
+		req.setPort(this.remotePort);
+
+		let bytes = req.serializeBinary();
+		writer.write(Buffer.from(bytes));
+	}
+
+	override toString() {
+		return `${super.toString()}`;
+	}
+}
+
+export class SupervisorSSHTunnel {
+
+	constructor(
+		private readonly logger: ILogger,
+		readonly workspaceInfo: WorkspaceAuthInfo,
+	) { }
+
+	private createPrivateKey(): Promise<string> {
+		const metadata = new BrowserHeaders();
+		metadata.append('x-gitpod-owner-token', this.workspaceInfo.ownerToken);
+		return new Promise((resolve, reject) => {
+			grpc.unary(ControlService.CreateSSHKeyPair, {
+				request: new CreateSSHKeyPairRequest(),
+				host: `${this.workspaceUrl}/_supervisor/v1`,
+				metadata,
+				onEnd: res => {
+					const { status, statusMessage, message } = res;
+					if (status === grpc.Code.OK && message instanceof CreateSSHKeyPairResponse) {
+						resolve(message.toObject().privateKey);
+					} else {
+						reject(statusMessage);
+					}
+				}
+			});
+		});
+	}
+
+	private get workspaceUrl() {
+		return `https://${this.workspaceInfo.workspaceId}.${this.workspaceInfo.workspaceHost}`;
+	}
+
+	private get workspaceWSUrl() {
+		return `wss://${this.workspaceInfo.workspaceId}.${this.workspaceInfo.workspaceHost}`;
+	}
+
+	public async establishTunnel(): Promise<ConnectConfig> {
```

Comment:
Could you map out analytics funnel how we will track conversion from hadnling uri to getting the session please. We should make sure to install all necessary analytics events and put GCP error reporting in such points.

Reply:
See `Analysis Path` section in PR description